qa/L0_long_running_stress/stress.py

```diff
@@ -0,0 +1,612 @@
+        finally:
+            _thread_exceptions_mutex.release()
+
+    # We need to explicitly close each client so that streams get
+    # cleaned up and closed correctly, otherwise the application
+    # can hang when exiting.
+    for c, i in client_metadata_list:
+        print("thread {} closing client {}".format(name, i))
+        c.close()
+
+    print("Exiting thread {}".format(name))
+
+
+def check_status(model_name):
+    client = grpcclient.InferenceServerClient("localhost:8001",
+                                              verbose=FLAGS.verbose)
+    stats = client.get_inference_statistics(model_name)
+    print(stats)
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+    parser.add_argument('-v',
+                        '--verbose',
+                        action="store_true",
+                        required=False,
+                        default=False,
+                        help='Enable verbose output')
+    parser.add_argument('-r',
+                        '--random-seed',
+                        type=int,
+                        required=False,
+                        help='Random seed.')
+    parser.add_argument('-t',
+                        '--concurrency',
+                        type=int,
+                        required=False,
+                        default=8,
+                        help='Request concurrency. Default is 8.')
+    parser.add_argument(
+        '-i',
+        '--iterations',
+        type=int,
+        required=False,
+        default=200,
+        help='Number of iterations of stress test to run. Default is 200.')
+    FLAGS = parser.parse_args()
+
+    # Initialize the random seed. For reproducibility each thread
+    # maintains its own RNG which is initialized based on this seed.
+    randseed = 0
+    if FLAGS.random_seed != None:
+        randseed = FLAGS.random_seed
+    else:
+        randseed = int(time.time())
+    np.random.seed(randseed)
+
+    print("random seed = {}".format(randseed))
+    print("concurrency = {}".format(FLAGS.concurrency))
+    print("iterations = {}".format(FLAGS.iterations))
+
+    for TARGET in BACKENDS.split():
+        trial = TARGET
+        dtype = get_datatype(trial)
+        model_name = tu.get_sequence_model_name(trial, dtype)
+
+        threads = []
+        for idx, thd in enumerate(range(FLAGS.concurrency)):
+            thread_name = "thread_{}".format(idx)
+
+            # Create the seed for the thread. Since these are created in
+            # reproducible order off of the initial seed we will get
+            # reproducible results when given the same seed.
+            seed = np.random.randint(2**32)
+
+            # Each thread is reserved a block of correlation IDs or size
+            # CORRELATION_ID_BLOCK_SIZE
+            correlation_id_base = 1 + (idx * CORRELATION_ID_BLOCK_SIZE)
+
+            threads.append(
+                threading.Thread(target=stress_thread,
```

Comment:
You should expand stress_thread to have more cases to cover different types of tests. For example, add some cases for timeouts, some cases for batched and non-batched requests (using nomodels but also using some real models like resnet). For misbehaving clients that crash you will need to run separate python scripts. But I would launch them as a separate case in stress_thread. You can then launch some clients that have bad behavior (like close connection, exit early, etc.). You can also launch some clients that start a transaction and then this script can kill them, etc. In general everything should be driven from this script.

Reply:
Will modify.